src/cockpit/samples.py

```diff
@@ -301,6 +301,60 @@ def sample(self, samples: Samples) -> None:
                     samples['cgroup.cpu.usage'][cgroup] = usage_nsec / 1000000
 
 
+class CGroupDiskIO(Sampler):
+    descriptions = [
+        SampleDescription('disk.cgroup.read', 'bytes', 'counter', True),
+        SampleDescription('disk.cgroup.written', 'bytes', 'counter', True),
+    ]
+
+    @staticmethod
+    def get_cgroup_name(fd: int) -> Optional[str]:
+        cgroup_name = ""
+        try:
+            with Handle.open('cgroup', os.O_RDONLY, dir_fd=fd) as cgroup_fd:
+                cgroup_name = os.read(cgroup_fd, 1024).decode().strip()
+        except FileNotFoundError:
+            return None
+
+        # Skip leadin ::0/
+        return cgroup_name[4:]
+
+    @staticmethod
+    def get_proc_io(fd: int) -> Optional[List[int]]:
+        data = ""
+        try:
+            with Handle.open('io', os.O_RDONLY, dir_fd=fd) as io_fd:
+                data = os.read(io_fd, 1024).decode()
+        except (FileNotFoundError, PermissionError):
+            return None
+
+        values = []
+        for key in ['read_bytes:', 'write_bytes:']:
+            start = data.index(key) + len(key)
+            end = data.index('\n', start)
+            values.append(int(data[start:end]))
+
+        return values
+
+    def sample(self, samples: Samples):
+        with Handle.open('/proc', os.O_RDONLY | os.O_DIRECTORY) as proc_fd:
+            for path in os.listdir(proc_fd):
+                # non-pid entries in proc are guaranteed to start with a character a-z
+                if path[0] < '0' or path[0] > '9':
+                    continue
+
+                try:
+                    with Handle.open(path, os.O_RDONLY | os.O_DIRECTORY, dir_fd=proc_fd) as pid_fd:
+                        cgroup_name = self.get_cgroup_name(pid_fd)
+                        proc_io = self.get_proc_io(pid_fd)
+
+                        if proc_io is not None:
+                            samples['disk.cgroup.read'][cgroup_name] = samples['disk.cgroup.read'].get(cgroup_name, 0) + proc_io[0]
```

Comment:
I think the best option here is to capture `samples['disk.cgroup.read']` and `...written...` at the top of the loop and use them here to avoid having to say `samples['disk.cgroup.read']` every time, twice.